### PR TITLE
Feat/token acq

### PR DESCRIPTION
**Pull request by royscheeren**
# Scope

Allow the access control query to handle token balances

# Work done

Add an option for token balance to the access control query

This PR includes a refactor of the xtz balance access control query.

### CODE REVIEW

File: src/data/data.test.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            },
+            value: 'VALUE',
+          },
+        ],
+      ],
+    ])('should get the ledger data from the storage of a contract', async (storage, expected) => {
+      // when ... we want the ledger from the storage of a contract
+      // then ... it should fetch and format as expected
+      const httpStub = { get: jest.fn().mockResolvedValue(storage) }
+      const result = await SUT._getLedgerFromStorage(httpStub as any)({
```
Comment (jeroenbranje) [inline src/data/data.test.ts:52]: Can we give the correct type here instead of any?

File: src/data/data.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    http
+      .get(
+        `https://${API_URLS[network]}/v1/tokens/balances?account.eq=${pkh}&token.contract.eq=${contract}&token.tokenId.eq=${tokenId}`,
+      )
+      .then(
+        pipe(
+          pathOr('0', ['data', 0]),
+          paths([['metadata', 'decimals'], ['balance']]) as (obj: any) => [string, string],
+          map(parseInt),
+          ([x, y]: number[]) => divide(y, 10 ** x),
```
Comment (jeroenbranje) [inline src/data/data.ts]: You can make a utility of this denominate function and test it

File: src/siwt.ts [hunk 3]
```diff
@@ -7,24 +7,25 @@
...
-import { ACCESS_TOKEN_EXPIRATION, API_URLS, ID_TOKEN_EXPIRATION, REFRESH_TOKEN_EXPIRATION } from './constants'
-import {
-  filterOwnedAssets,
-  getOwnedAssetIds,
-} from './utils/siwt.utils'
+import { ACCESS_TOKEN_EXPIRATION, ID_TOKEN_EXPIRATION, REFRESH_TOKEN_EXPIRATION } from './constants'
+import { validateNFTCondition, validateTokenBalanceCondition, validateXTZBalanceCondition } from './utils/siwt.utils'
+import { getBalance, getLedgerFromStorage } from './data'
+import { http } from './http'
+import { getTokenBalance } from './data/data'
```
Comment (jeroenbranje) [inline src/siwt.ts]: import from data/data lets define it in the index

File: src/utils/siwt.utils.ts [hunk 4]
```diff
@@ -54,7 +62,7 @@ export const packMessagePayload = (messageData: MessagePayloadData): string =>
     join(''),
   )(messageData)
 
-export const filterOwnedAssetsFromNFTAssetContract = (pkh: string) => filter(propEq('value', pkh))
+export const filterOwnedAssetsFromNFTAssetContract = (pkh: string) => pipe(filter(propEq('value', pkh)))
```
Comment (jeroenbranje) [inline src/utils/siwt.utils.ts]: Is pipe necessary here?

File: src/utils/siwt.utils.ts [hunk 5]
```diff
@@ -70,9 +78,18 @@ export const determineContractAssetType = pipe(
 
 export const filterOwnedAssets = (pkh: string) =>
   cond([
-    [pipe(determineContractAssetType, equals(AssetContractType.nft)), filterOwnedAssetsFromNFTAssetContract(pkh) as any],
-    [pipe(determineContractAssetType, equals(AssetContractType.multi)), filterOwnedAssetsFromMultiAssetContract(pkh) as any],
-    [pipe(determineContractAssetType, equals(AssetContractType.single)), filterOwnedAssetsFromSingleAssetContract(pkh) as any],
+    [
+      pipe(determineContractAssetType, equals(AssetContractType.nft)),
+      filterOwnedAssetsFromNFTAssetContract(pkh) as any,
```
Comment (jeroenbranje) [inline src/utils/siwt.utils.ts]: any type?

File: src/utils/siwt.utils.ts [hunk 6]
```diff
@@ -82,3 +99,55 @@ export const getOwnedAssetIds = cond([
...
+    parameters: { pkh },
+    test: { contractAddress, comparator, value },
+  }: AccessControlQuery) =>
+    getLedgerFromStorage({ network, contract: contractAddress })
+      .then(storage => {
+        const ownedAssets = filterOwnedAssets(pkh as string)(storage as LedgerStorage[])
+        const ownedAssetIds = getOwnedAssetIds(ownedAssets)
+
+        return {
+          passed: (COMPARISONS[comparator] as any)(prop('length')(ownedAssets))(value),
```
Comment (jeroenbranje) [inline src/utils/siwt.utils.ts:129]: Is this any not an "Comparator" type

**Review by jeroenbranje — CHANGES_REQUESTED** [review]

File: src/data/data.test.ts (hunk 1 above)
Comment (royscheeren) [reply src/data/data.test.ts:52]: no, its a test

**Review by jeroenbranje — DISMISSED** [review]

**Review by jeroenbranje — APPROVED** [review]

**Pull request merged**